dynamic_instrumentation/src/instrument.rs

```diff
@@ -123,6 +129,198 @@ fn is_region_or_unsafe_ptr(ty: Ty) -> bool {
+}
+
+impl<'tcx> MutVisitor<'tcx> for RewriteAddressTakenLocals<'tcx> {
+    /// Rewrites an address-taken local in terms of its underlying address.
+    fn visit_place(
+        &mut self,
+        mut place: &mut Place<'tcx>,
+        context: PlaceContext,
+        location: Location,
+    ) {
+        // If we have found an address-taken local `_x`, substitute with `(*_y)` where `_y`
+        // is the address of `_x`.
+        if let Some(substitute) = self.local_substitute.get(&place.local) {
+            if context.is_use()
+                && (!context.is_place_assignment()
+                    || (context.is_place_assignment()
+                        && !place.is_indirect()
+                        && place.projection.len() > 0))
+                && !context.is_drop()
+                && !matches!(
+                    context,
+                    PlaceContext::NonMutatingUse(NonMutatingUseContext::Inspect)
+                )
+            {
+                // add deref
+                let projection = {
+                    let mut v = Vec::with_capacity(place.projection.len() + 1);
+                    v.extend([ProjectionElem::Deref]);
+                    v.extend(place.projection);
+                    self.tcx().intern_place_elems(&v)
+                };
+                // replace `_x` with `_y`
+                place.local = *substitute;
+                place.projection = projection;
+            }
+        }
+
+        self.super_place(place, context, location)
+    }
+
+    fn tcx<'a>(&'a self) -> TyCtxt<'tcx> {
+        self.tcx()
+    }
+
+    fn visit_body(&mut self, body: &mut Body<'tcx>) {
+        let mut local_address_statements = Vec::new();
+
+        // for each address-taken local, push a new local that will get assigned
+        // its address
+        for local in &self.address_taken {
+            let arg_ty = Place::from(*local).ty(&body.local_decls, self.tcx()).ty;
+            let inner_ty = ty::TypeAndMut {
+                ty: arg_ty,
+                mutbl: Mutability::Mut,
+            };
+
+            let raw_ptr_ty = self.tcx().mk_ptr(inner_ty);
+            let raw_ptr_local = body.local_decls.push(LocalDecl::new(raw_ptr_ty, DUMMY_SP));
+
+            self.local_substitute.insert(*local, raw_ptr_local);
+
+            // set original local as mutable
+            body.local_decls.get_mut(*local).unwrap().mutability = Mutability::Mut;
+        }
+
+        // START ADDRESS-TAKING STATEMENTS
+        //
+        // The code below appends context-dependent MIR locations to a collection
+        // that will later be used as reference for where to insert the first instance
+        // of taking the address of an already-determined-to-be address-taken local.
+        // E.g. if `_x` is an address-taken local, below is what determines where the
+        // statement `_y = &raw _x` will be placed.
+        // assign insertion location of address-taken arguments
+        for l in 1..body.arg_count + 1 {
```

Comment:
```suggestion
        for l in 1..=body.arg_count {
```

Reply:
`0` is the return local so we want to ignore it, i'll take your suggestion